src/Headers/Ic.h

```diff
@@ -1043,42 +1046,118 @@ class RegularizerFunction {
   virtual ~RegularizerFunction() { } ;
 
   virtual FLOAT operator()(const Particle<ndim>&, const Particle<ndim>&) const  = 0 ;
+  //virtual FLOAT GetSmoothedDensity(const FLOAT *, const FLOAT) = 0;
 };
 
+
 // Standard regularizer based upon density and particle disorder
 template<int ndim, class DensityFunc>
-class DefaultRegularizerFunction : public RegularizerFunction<ndim> {
-  FLOAT alphaReg, rhoReg ;
-  const DensityFunc* _rho_func ;
+class DefaultRegularizerFunction : public RegularizerFunction<ndim>
+{
+  const bool regularise_smooth_density;
+  const FLOAT alphaReg, rhoReg;
+  const DensityFunc* _rho_func;
+  SmoothingKernel<ndim>* _kern;
+
 public:
-  DefaultRegularizerFunction(Parameters* simparams, const DensityFunc* rho)
-  : alphaReg(simparams->floatparams["alpha_reg"]),
-    rhoReg  (simparams->floatparams["rho_reg"]),
-    _rho_func(rho)
+  DefaultRegularizerFunction(SmoothingKernel<ndim> *kern, Parameters* simparams, const DensityFunc* rho)
+  : regularise_smooth_density(simparams->intparams["regularise_smooth_density"]),
+    alphaReg(simparams->floatparams["alpha_reg"]),
+    rhoReg(simparams->floatparams["rho_reg"]),
+    _rho_func(rho),
+    _kern(kern)
 { } ;
 
   virtual ~DefaultRegularizerFunction() { } ;
 
+
   virtual FLOAT operator()(const Particle<ndim>& part,
                            const Particle<ndim>& neibpart) const
   {
-    FLOAT rhotrue = _rho_func->GetDensity(neibpart.r);
+    FLOAT rhotrue;
+
+    if (regularise_smooth_density) {
```

Comment:
Can we not just have two different RegularizerFunction function objects? One a DefaultRegularizerFunction and one a SmoothedRegularizerFunction? 

This would be in better agreement with the pricinciples of OOP. I'm happy to do implementation this if desired...